components/script/dom/globalscope.rs

```diff
@@ -207,6 +378,181 @@ impl GlobalScope {
         }
     }
 
+    pub fn mark_port_as_shipped(
+        &self,
+        port_id: MessagePortId,
+        entangled_port_id: Option<MessagePortId>,
+    ) {
+        self.message_ports_shipped
+            .borrow_mut()
+            .insert(port_id, entangled_port_id);
+    }
+
+    /// Route the task to be handled by the relevant port.
+    /// Note that the task is run while message_ports is borrowed.
+    /// Therefore, pending_message_ports should be used to track ports,
+    /// that are either transfer-received or created in a task.
+    pub fn route_task_to_port(&self, port_id: MessagePortId, task: PortMessageTask) {
+        if self
+            .message_ports_shipped
+            .borrow_mut()
+            .contains_key(&port_id)
+        {
+            let _ = self
+                .script_to_constellation_chan()
+                .send(ScriptMsg::RerouteMessagePort(port_id.clone(), task));
+            return;
+        }
+        if let Some(port) = self.pending_message_ports.borrow_mut().remove(&port_id) {
+            self.message_ports
+                .borrow_mut()
+                .insert(port_id.clone(), &port);
+        }
+        if let Some(port) = self.message_ports.borrow().find_port(&port_id) {
+            port.handle_incoming(task);
+        }
+    }
+
+    pub fn maybe_add_pending_ports(&self) {
+        // Add pending ports to message-ports.
+        for (port_id, port) in self.pending_message_ports.borrow_mut().drain_ports() {
+            if !self.message_ports_shipped.borrow().contains_key(&port_id) {
+                self.message_ports
+                    .borrow_mut()
+                    .insert(port_id.clone(), &port);
+                let control_sender = self
+                    .port_senders
+                    .borrow_mut()
+                    .remove(&port_id)
+                    .expect("This global to be tracking this port");
+                let _ = self
+                    .script_to_constellation_chan()
+                    .send(ScriptMsg::NewMessagePort(port_id.clone(), control_sender));
+            }
+        }
+    }
+
+    /// Mark a message-port as eligible for garbage collection.
+    pub fn start_garbage_collecting_message_port(&self, port_id: &MessagePortId) {
+        if self
+            .message_ports_to_be_collected
+            .borrow()
+            .contains_key(&port_id)
+        {
+            return;
+        }
+        self.message_ports_to_be_collected
+            .borrow_mut()
+            .insert(port_id.clone(), 0);
+    }
+
+    /// After a port has been marked as up for collection,
+    /// we wait two turns of the event-loop,
```

Comment:
Note: two "turns" is not equal to two tasks, since each turn drains all tasks currently enqueued. So it's two turns of the `select` of a script-thread or worker event-loop, which can end-up running more than two tasks.